### DISCUSSION TITLE
Caveat in listing the loaded shared libraries by adapting the `ctypes.util.find_library` CPython patch?

### DISCUSSION THREAD

**Discussion by lesteve** [pyodide/pyodide#4524] [Q&A]
So in [threadpoolctl](https://github.com/joblib/threadpoolctl) we want to inspect loaded shared libraries to put some info about BLAS and OpenMP mostly I think. cc @jeremiedbb who is the main maintainer.

Currently we get a warning inside Pyodide because threadpoolctl relies on finding `libc` and using `dl_iterate_phdr`. That does not work inside Pyodide because `ctypes.util.find_library('c')` returns `[]`. See https://github.com/joblib/threadpoolctl/issues/167 for more details.

Adapting the logic in the Pyodide `ctypes.util.find_library` in 
https://github.com/pyodide/pyodide/blob/main/cpython/patches/0003-Add-emscripten-platform-support-to-ctypes.util.find_.patch?rgh-link-date=2024-02-14T16%3A52%3A08Z, I have come up with the following implementation:
```py
def list_libraries():
    libraries = []
    ld_library_paths = [
        path
        for path in os.environ.get("LD_LIBRARY_PATH", "").split(":")
        if os.path.exists(path)
    ]
    for dir in ld_library_paths:
        dir_full_paths = [os.path.join(dir, name) for name in os.listdir(dir)]
        dir_libraries = [
            path
            for path in dir_full_paths
            if path.endswith(".so") or path.endswith(".wasm") and _is_wasm(path)
        ]
        libraries += dir_libraries
    return libraries
```

Is there any caveat in using this to detect loaded shared libraries in Pyodide?

Also out of curiosity, what kind of libraries have a `.wasm` extension?

My current PR in threadpoolctl is https://github.com/joblib/threadpoolctl/pull/169 if you are curious :wink:.

**Comment by hoodmane** [comment]
Thanks @lesteve. I'll reply in the threadpoolctl issue.

> Also out of curiosity, what kind of libraries have a .wasm extension?

It's unusual, but there are some .wasm files floating around. @ryanking13 do you know of any examples that do this? Or was this a just in case thing?

**Reply by ryanking13** [reply]
> It's unusual, but there are some .wasm files floating around. @ryanking13 do you know of any examples that do this? Or was this a just in case thing?

It was mostly just in case thing.

**Comment by hoodmane** [comment]
I'll close this as resolved since I think we figured this out. Thanks @lesteve!
